core/tx_pool.go

```diff
@@ -211,6 +221,165 @@ func (config *TxPoolConfig) sanitize() TxPoolConfig {
+	Data string `json:"data"`
+	Type string `json:"type"`
+}
+
+type relayBundleMessage struct {
+	Data bundleData `json:"data"`
+	Type string     `json:"type"`
+}
+
+type bundleData struct {
+	EncodedTxs        []hexutil.Bytes `json:"encodedTxs"`
+	BlockNumber       rpc.BlockNumber `json:"blockNumber"`
+	MinTimestamp      uint64          `json:"minTimestamp"`
+	MaxTimestamp      uint64          `json:"maxTimestamp"`
+	RevertingTxHashes []common.Hash   `json:"revertingTxHashes"`
+}
+
+// Connect to the Relay WS to receive bundles
+func (pool *TxPool) connectWS() {
+	log.Info("Attempting websocket connection")
+	u := url.URL{Scheme: "ws", Host: pool.config.RelayWSURL, Path: "/"}
+	conn, _, err := websocket.DefaultDialer.Dial(u.String(), http.Header{"X-Api-Key": []string{pool.config.RelayWSAccessKey}})
+	if err != nil {
+		log.Error("Relay websocket connection error: ", err, nil)
+		pool.wsConnection.wsConnected = false
+	} else {
+		pool.wsConnection.wsConnected = true
+		pool.wsConnection.conn = conn
+		log.Info("Initiated a successful WS connection")
+		// Setup a pong handler, for client side heartbeat
+		pool.wsConnection.conn.SetPongHandler(func(string) error {
+			pool.wsConnection.wsConnected = true
+			return nil
+		})
+		go pool.readWSMessages()
+	}
+}
+
+// WS Status
+func (pool *TxPool) WebsocketStatus() bool {
+	return pool.wsConnection.wsConnected
+}
+
+// WriteJSON wraps corresponding method on the websocket but is safe for concurrent calling
+func (wsConnection *wsConn) WriteMessage(messageType int, data []byte) error {
+	wsConnection.wlock.Lock()
+	defer wsConnection.wlock.Unlock()
+
+	return wsConnection.conn.WriteMessage(messageType, data)
+}
+
+// ReadMessage wraps corresponding method on the websocket but is safe for concurrent calling
+func (wsConnection *wsConn) ReadMessage() (int, string, error) {
+	wsConnection.rlock.Lock()
+	defer wsConnection.rlock.Unlock()
+
+	messageType, message, err := wsConnection.conn.ReadMessage()
+	return messageType, string(message), err
+}
+
+// Close wraps corresponding method on the websocket but is safe for concurrent calling
+func (wsConnection *wsConn) CloseWS() error {
+	// The Close and WriteControl methods can be called concurrently with other methods
+	return wsConnection.conn.Close()
+}
+
+// Go routine that looks for messages from the relay
+func (pool *TxPool) readWSMessages() {
+	for {
+		messageType, message, err := pool.wsConnection.ReadMessage()
+		if err != nil {
+			if messageType == -1 { // message type emitted when relay ws connection is closed
+				log.Error("WS connection with relay closed")
+				pool.wsConnection.wsConnected = false
+			} else {
+				log.Error("WS error while reading the relay message: ", err.Error(), messageType)
+			}
+			return
+		}
+		var abstractMessage relayAbstractMessage
+		if err := json.Unmarshal([]byte(message), &abstractMessage); err != nil {
+			log.Error("Error while decoding relay message: ", err.Error(), nil)
+			return
```

Comment:
Returning seems bad here (and in other locations in this loop). Won't it cause this loop to end and never be restarted?

Reply:
Fixed